<a href="https://colab.research.google.com/github/Bryan-Benisti/Data_Analyse_projects/blob/main/Projet%202_UFC_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#BRYAN BENISTI

In [ ]:
#First I import the Libraries

import pandas as pd      

import seaborn as sns    

import matplotlib.pyplot as plt

from datetime import datetime

import statsmodels.formula.api as smf 

In [ ]:
#from google.colab import drive

#drive.mount('/content/drive')

In [ ]:
#Then I Load the file from my Google Drive

path = "https://drive.google.com/file/d/1xD_7QuYBZWw4F_e5rSC8-4miktPim-ms/view?usp=sharing"

path ="https://drive.google.com/uc?id=" + path.split('/')[-2]

My_Data = pd.read_csv(path)

In [ ]:
#Now let's observe it a bit

#Here are the first 3 lines 

My_Data.head(3)

Event Name      Event Date          Result  \
0  UFC Fight Night: Vera vs. Sandhagen  March 25, 2023  Cory Sandhagen   
1  UFC Fight Night: Vera vs. Sandhagen  March 25, 2023      Holly Holm   
2  UFC Fight Night: Vera vs. Sandhagen  March 25, 2023   Nate Landwehr   

         Fighter1      Fighter2   KD Strikes   TD  Sub          Weight Class  \
0  Cory Sandhagen   Marlon Vera  0-0  128-58  3-0  0-1          Bantamweight   
1      Holly Holm   Yana Santos  0-0   32-21  4-0  0-0  Women's Bantamweight   
2   Nate Landwehr  Austin Lingo  0-0   64-48  1-0  1-0         Featherweight   

                 Method  Round  Time  
0                 S-DEC      5  5:00  
1                 U-DEC      3  5:00  
2  SUB-Rear-Naked-Choke      2  4:11

In [ ]:
#Here we can observe a list of all the columns 

df_columns = list(My_Data.columns)

df_columns

['Event Name',
 'Event Date',
 'Result',
 'Fighter1',
 'Fighter2',
 'KD',
 'Strikes',
 'TD',
 'Sub',
 'Weight Class',
 'Method',
 'Round',
 'Time']

In [ ]:
#Here we can have informations about the variables on the dataframe 

My_Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7060 entries, 0 to 7059
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Event Name    7060 non-null   object
 1   Event Date    7060 non-null   object
 2   Result        7060 non-null   object
 3   Fighter1      7060 non-null   object
 4   Fighter2      7060 non-null   object
 5   KD            7060 non-null   object
 6   Strikes       7060 non-null   object
 7   TD            7060 non-null   object
 8   Sub           7060 non-null   object
 9   Weight Class  7060 non-null   object
 10  Method        7060 non-null   object
 11  Round         7060 non-null   int64 
 12  Time          7060 non-null   object
dtypes: int64(1), object(12)
memory usage: 717.2+ KB


In [ ]:
#Let's have some descriptive statistics about all the columns 

My_Data.describe(include="all")

Event Name         Event Date   Result    Fighter1  \
count                7060               7060     7060        7060   
unique                640                635     1629        1638   
top     UFC 2: No Way Out  November 19, 2016  Unknown  Jim Miller   
freq                   15                 25       71          25   
mean                  NaN                NaN      NaN         NaN   
std                   NaN                NaN      NaN         NaN   
min                   NaN                NaN      NaN         NaN   
25%                   NaN                NaN      NaN         NaN   
50%                   NaN                NaN      NaN         NaN   
75%                   NaN                NaN      NaN         NaN   
max                   NaN                NaN      NaN         NaN   

               Fighter2    KD Strikes    TD   Sub Weight Class Method  \
count              7060  7060    7060  7060  7060         7060   7060   
unique             2245    16    3618    82    47           15     71   
top     Jeremy Stephens   0-0     0-0   0-0   0-0  Lightweight  U-DEC   
freq                 19  4559      29  1891  3957         1237   2464   
mean                NaN   NaN     NaN   NaN   NaN          NaN    NaN   
std                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   
min                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   
25%                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   
50%                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   
75%                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   
max                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   

              Round  Time  
count   7060.000000  7060  
unique          NaN   339  
top             NaN  5:00  
freq            NaN  3280  
mean       2.332011   NaN  
std        1.014400   NaN  
min        1.000000   NaN  
25%        1.000000   NaN  
50%        3.000000   NaN  
75%        3.000000   NaN  
max        5.000000   NaN

In [ ]:
# Lets create a new DataFrame by filtering the original one

# First I create a New column called DateYear that gives me only the Year of the events dates

My_Data['DateYear'] = pd.to_datetime(My_Data['Event Date']).dt.year

My_Data.describe(include="all")

Event Name         Event Date   Result    Fighter1  \
count                7060               7060     7060        7060   
unique                640                635     1629        1638   
top     UFC 2: No Way Out  November 19, 2016  Unknown  Jim Miller   
freq                   15                 25       71          25   
mean                  NaN                NaN      NaN         NaN   
std                   NaN                NaN      NaN         NaN   
min                   NaN                NaN      NaN         NaN   
25%                   NaN                NaN      NaN         NaN   
50%                   NaN                NaN      NaN         NaN   
75%                   NaN                NaN      NaN         NaN   
max                   NaN                NaN      NaN         NaN   

               Fighter2    KD Strikes    TD   Sub Weight Class Method  \
count              7060  7060    7060  7060  7060         7060   7060   
unique             2245    16    3618    82    47           15     71   
top     Jeremy Stephens   0-0     0-0   0-0   0-0  Lightweight  U-DEC   
freq                 19  4559      29  1891  3957         1237   2464   
mean                NaN   NaN     NaN   NaN   NaN          NaN    NaN   
std                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   
min                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   
25%                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   
50%                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   
75%                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   
max                 NaN   NaN     NaN   NaN   NaN          NaN    NaN   

              Round  Time     DateYear  
count   7060.000000  7060  7060.000000  
unique          NaN   339          NaN  
top             NaN  5:00          NaN  
freq            NaN  3280          NaN  
mean       2.332011   NaN  2014.545042  
std        1.014400   NaN     6.017996  
min        1.000000   NaN  1994.000000  
25%        1.000000   NaN  2011.000000  
50%        3.000000   NaN  2016.000000  
75%        3.000000   NaN  2019.000000  
max        5.000000   NaN  2023.000000

In [ ]:
#Now let's filter the data frame keeping only the data between 2022 to 2023 using our new column

My_Data_2022_to_2023 = My_Data[(My_Data['DateYear'] >= 2022) & (My_Data['DateYear'] <= 2023)]

#let's see the difference of rows'number

print(My_Data.shape)

print(My_Data_2022_to_2023.shape)

(7060, 14)
(634, 14)


In [ ]:
# Now I want to filter again but this time I want to keep only the main event of each dates from 2022 to now. 

Main_Fight = My_Data_2022_to_2023["Event Date"].unique()  

My_Data_Mains_Events = pd.DataFrame() 

for date in Main_Fight:

    fight = My_Data_2022_to_2023.loc[My_Data_2022_to_2023["Event Date"] == date].iloc[0]  

    My_Data_Mains_Events = My_Data_Mains_Events.append(fight, ignore_index=True)  

My_Data_Mains_Events.head(51)



<ipython-input-12-72e34c3bc4c0>:11: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  My_Data_Mains_Events = My_Data_Mains_Events.append(fight, ignore_index=True)


Event Name          Event Date  \
0          UFC Fight Night: Vera vs. Sandhagen      March 25, 2023   
1                 UFC 286: Edwards vs. Usman 3      March 18, 2023   
2         UFC Fight Night: Yan vs. Dvalishvili      March 11, 2023   
3                      UFC 285: Jones vs. Gane      March 04, 2023   
4             UFC Fight Night: Muniz vs. Allen   February 25, 2023   
5     UFC Fight Night: Andrade vs. Blanchfield   February 18, 2023   
6           UFC 284: Makhachev vs. Volkanovski   February 11, 2023   
7            UFC Fight Night: Lewis vs. Spivac   February 04, 2023   
8                   UFC 283: Teixeira vs. Hill    January 21, 2023   
9       UFC Fight Night: Strickland vs. Imavov    January 14, 2023   
10   UFC Fight Night: Cannonier vs. Strickland   December 17, 2022   
11            UFC 282: Blachowicz vs. Ankalaev   December 10, 2022   
12       UFC Fight Night: Thompson vs. Holland   December 03, 2022   
13     UFC Fight Night: Nzechukwu vs. Cutelaba   November 19, 2022   
14               UFC 281: Adesanya vs. Pereira   November 12, 2022   
15        UFC Fight Night: Rodriguez vs. Lemos   November 05, 2022   
16           UFC Fight Night: Kattar vs. Allen    October 29, 2022   
17             UFC 280: Oliveira vs. Makhachev    October 22, 2022   
18          UFC Fight Night: Grasso vs. Araujo    October 15, 2022   
19               UFC Fight Night: Dern vs. Yan    October 01, 2022   
20         UFC Fight Night: Sandhagen vs. Song  September 17, 2022   
21                  UFC 279: Diaz vs. Ferguson  September 10, 2022   
22           UFC Fight Night: Gane vs. Tuivasa  September 03, 2022   
23                  UFC 278: Usman vs. Edwards     August 20, 2022   
24              UFC Fight Night: Vera vs. Cruz     August 13, 2022   
25            UFC Fight Night: Santos vs. Hill     August 06, 2022   
26                   UFC 277: Pena vs. Nunes 2       July 30, 2022   
27       UFC Fight Night: Blaydes vs. Aspinall       July 23, 2022   
28       UFC Fight Night: Ortega vs. Rodriguez       July 16, 2022   
29       UFC Fight Night: Dos Anjos vs. Fiziev       July 09, 2022   
30             UFC 276: Adesanya vs. Cannonier       July 02, 2022   
31       UFC Fight Night: Tsarukyan vs. Gamrot       June 25, 2022   
32          UFC Fight Night: Kattar vs. Emmett       June 18, 2022   
33             UFC 275: Teixeira vs. Prochazka       June 11, 2022   
34     UFC Fight Night: Volkov vs. Rozenstruik       June 04, 2022   
35            UFC Fight Night: Holm vs. Vieira        May 21, 2022   
36       UFC Fight Night: Blachowicz vs. Rakic        May 14, 2022   
37               UFC 274: Oliveira vs. Gaethje        May 07, 2022   
38              UFC Fight Night: Font vs. Vera      April 30, 2022   
39          UFC Fight Night: Lemos vs. Andrade      April 23, 2022   
40         UFC Fight Night: Luque vs. Muhammad      April 16, 2022   
41  UFC 273: Volkanovski vs. The Korean Zombie      April 09, 2022   
42        UFC Fight Night: Blaydes vs. Daukaus      March 26, 2022   
43        UFC Fight Night: Volkov vs. Aspinall      March 19, 2022   
44        UFC Fight Night: Santos vs. Ankalaev      March 12, 2022   
45             UFC 272: Covington vs. Masvidal      March 05, 2022   
46        UFC Fight Night: Makhachev vs. Green   February 26, 2022   
47            UFC Fight Night: Walker vs. Hill   February 19, 2022   
48           UFC 271: Adesanya vs. Whittaker 2   February 12, 2022   
49  UFC Fight Night: Hermansson vs. Strickland   February 05, 2022   
50                   UFC 270: Ngannou vs. Gane    January 22, 2022   

                   Result               Fighter1               Fighter2   KD  \
0          Cory Sandhagen         Cory Sandhagen            Marlon Vera  0-0   
1            Leon Edwards           Leon Edwards           Kamaru Usman  0-0   
2       Merab Dvalishvili      Merab Dvalishvili               Petr Yan  0-0   
3               Jon Jones              Jon Jones             Ciryl Gane  0-

In [ ]:
# Number of fights in each event:

Nb_of_fight_grouped = My_Data_2022_to_2023.groupby("Event Name",as_index=False).agg({"Result": 'count'})

Nb_of_fight_grouped


Event Name  Result
0                    UFC 270: Ngannou vs. Gane      11
1            UFC 271: Adesanya vs. Whittaker 2      14
2              UFC 272: Covington vs. Masvidal      13
3   UFC 273: Volkanovski vs. The Korean Zombie      12
4                UFC 274: Oliveira vs. Gaethje      14
5              UFC 275: Teixeira vs. Prochazka      11
6              UFC 276: Adesanya vs. Cannonier      12
7                    UFC 277: Pena vs. Nunes 2      13
8                   UFC 278: Usman vs. Edwards      12
9                   UFC 279: Diaz vs. Ferguson      13
10             UFC 280: Oliveira vs. Makhachev      12
11               UFC 281: Adesanya vs. Pereira      14
12            UFC 282: Blachowicz vs. Ankalaev      12
13                  UFC 283: Teixeira vs. Hill      15
14          UFC 284: Makhachev vs. Volkanovski      13
15                     UFC 285: Jones vs. Gane      14
16                UFC 286: Edwards vs. Usman 3      15
17    UFC Fight Night: Andrade vs. Blanchfield      11
18       UFC Fight Night: Blachowicz vs. Rakic      11
19       UFC Fight Night: Blaydes vs. Aspinall      14
20        UFC Fight Night: Blaydes vs. Daukaus      12
21   UFC Fight Night: Cannonier vs. Strickland      13
22               UFC Fight Night: Dern vs. Yan      11
23       UFC Fight Night: Dos Anjos vs. Fiziev      11
24              UFC Fight Night: Font vs. Vera      11
25           UFC Fight Night: Gane vs. Tuivasa      12
26          UFC Fight Night: Grasso vs. Araujo      11
27  UFC Fight Night: Hermansson vs. Strickland      13
28            UFC Fight Night: Holm vs. Vieira      11
29           UFC Fight Night: Kattar vs. Allen      11
30        UFC Fight Night: Kattar vs. Chikadze      10
31          UFC Fight Night: Kattar vs. Emmett      13
32          UFC Fight Night: Lemos vs. Andrade      11
33           UFC Fight Night: Lewis vs. Spivac      11
34         UFC Fight Night: Luque vs. Muhammad      14
35        UFC Fight Night: Makhachev vs. Green      11
36            UFC Fight Night: Muniz vs. Allen      10
37     UFC Fight Night: Nzechukwu vs. Cutelaba      11
38       UFC Fight Night: Ortega vs. Rodriguez      12
39        UFC Fight Night: Rodriguez vs. Lemos      11
40         UFC Fight Night: Sandhagen vs. Song      13
41        UFC Fight Night: Santos vs. Ankalaev      14
42            UFC Fight Night: Santos vs. Hill      10
43      UFC Fight Night: Strickland vs. Imavov      11
44       UFC Fight Night: Thompson vs. Holland      14
45       UFC Fight Night: Tsarukyan vs. Gamrot      12
46              UFC Fight Night: Vera vs. Cruz      13
47         UFC Fight Night: Vera vs. Sandhagen      10
48        UFC Fight Night: Volkov vs. Aspinall      12
49     UFC Fight Night: Volkov vs. Rozenstruik      14
50            UFC Fight Night: Walker vs. Hill      12
51        UFC Fight Night: Yan vs. Dvalishvili      13

In [ ]:
# Average rounds by fighter

avg_rounds_won_by_fighter = My_Data_2022_to_2023.groupby("Result", as_index=False).agg({"Round": lambda x: sum([int(n) for n in x]) / len(x)})

avg_rounds_won_by_fighter


Result  Round
0              AJ Fletcher    2.0
1     Abdul Razak Alhassan    2.0
2    Abubakar Nurmagomedov    3.0
3           Abus Magomedov    1.0
4              Adam Fugitt    1.0
..                     ...    ...
417          Yanal Ashmouz    1.0
418        Yazmin Jauregui    2.5
419         Yohan Lainesse    3.0
420              Zac Pauga    3.0
421            Zhang Weili    2.0

[422 rows x 2 columns]

In [ ]:
# create a pivot table, that shows the average number of rounds fought in each weight class for each event, with 1 digit after the point,
# And replacing "NaN" By 0

pivot_table = My_Data_2022_to_2023.pivot_table(values='Round', index=['Event Name', 'Event Date'], columns='Weight Class', aggfunc='mean')

pivot_table = pivot_table.round(1)

pivot_table = pivot_table.fillna(0)

pivot_table


Weight Class                                                   Bantamweight  \
Event Name                                 Event Date                         
UFC 270: Ngannou vs. Gane                  January 22, 2022             2.3   
UFC 271: Adesanya vs. Whittaker 2          February 12, 2022            2.7   
UFC 272: Covington vs. Masvidal            March 05, 2022               0.0   
UFC 273: Volkanovski vs. The Korean Zombie April 09, 2022               4.0   
UFC 274: Oliveira vs. Gaethje              May 07, 2022                 3.0   
UFC 275: Teixeira vs. Prochazka            June 11, 2022                3.0   
UFC 276: Adesanya vs. Cannonier            July 02, 2022                2.0   
UFC 277: Pena vs. Nunes 2                  July 30, 2022                0.0   
UFC 278: Usman vs. Edwards                 August 20, 2022              3.0   
UFC 279: Diaz vs. Ferguson                 September 10, 2022           3.0   
UFC 280: Oliveira vs. Makhachev            October 22, 2022             2.5   
UFC 281: Adesanya vs. Pereira              November 12, 2022            2.0   
UFC 282: Blachowicz vs. Ankalaev           December 10, 2022            2.0   
UFC 283: Teixeira vs. Hill                 January 21, 2023             2.5   
UFC 284: Makhachev vs. Volkanovski         February 11, 2023            0.0   
UFC 285: Jones vs. Gane                    March 04, 2023               3.0   
UFC 286: Edwards vs. Usman 3               March 18, 2023               0.0   
UFC Fight Night: Andrade vs. Blanchfield   February 18, 2023            0.0   
UFC Fight Night: Blachowicz vs. Rakic      May 14, 2022                 3.0   
UFC Fight Night: Blaydes vs. Aspinall      July 23, 2022                0.0   
UFC Fight Night: Blaydes vs. Daukaus       March 26, 2022               2.0   
UFC Fight Night: Cannonier vs. Strickland  December 17, 2022            2.5   
UFC Fight Night: Dern vs. Yan              October 01, 2022             2.0   
UFC Fight Night: Dos Anjos vs. Fiziev      July 09, 2022                3.0   
UFC Fight Night: Font vs. Vera             April 30, 2022               5.0   
UFC Fight Night: Gane vs. Tuivasa          September 03, 2022           1.0   
UFC Fight Night: Grasso vs. Araujo         October 15, 2022             2.7   
UFC Fight Night: Hermansson vs. Strickland February 05, 2022            3.0   
UFC Fight Night: Holm vs. Vieira           May 21, 2022                 3.0   
UFC Fight Night: Kattar vs. Allen          October 29, 2022             1.0   
UFC Fight Night: Kattar vs. Chikadze       January 15, 2022             0.0   
UFC Fight Night: Kattar vs. Emmett         June 18, 2022                1.0   
UFC Fight Night: Lemos vs. Andrade         April 23, 2022               1.0   
UFC Fight Night: Lewis vs. Spivac          February 04, 2023            1.0   
UFC Fight Night: Luque vs. Muhammad        April 16, 2022               1.0   
UFC Fight Night: Makhachev vs. Green       February 26, 2022            0.0   
UFC Fight Night: Muniz vs. Allen           February 25, 2023            0.0   
UFC Fight Night: Nzechukwu vs. Cutelaba    November 19, 2022            3.0   
UFC Fight Night: Ortega vs. Rodriguez      July 16, 2022                2.0   
UFC Fight Night: Rodriguez vs. Lemos       November 05, 2022            2.0   
UFC Fight Night: Sandhagen vs. Song        September 17, 2022           3.5   
UFC Fight Night: Santos vs. Ankalaev       March 12, 2022               1.7   
UFC Fight Night: Santos vs. Hill           August 06, 2022              0.0   
UFC Fight Night: Strickland vs. Imavov     January 14, 2023             2.0   
UFC Fight Night: Thompson vs. Holland      December 03, 2022            0.0   
UFC Fight Night: Tsarukyan vs. Gamrot      June 25, 2022                2.3   
UFC Fight Night: Vera vs. Cruz             August 13, 2022              3.5   
UFC Fight Night: Vera vs. Sandhagen        March 25, 2023               5.0   
UFC Fight Night: Volkov vs. Aspinall       March 1

In [ ]:
# Let's now make a Linear Regression 

Lin_Reg = smf.ols(formula = "Round ~ Time" , data = My_Data_2022_to_2023).fit()

print(Lin_Reg.summary())

                            OLS Regression Results                            
Dep. Variable:                  Round   R-squared:                       0.717
Model:                            OLS   Adj. R-squared:                  0.587
Method:                 Least Squares   F-statistic:                     5.494
Date:                Sun, 16 Apr 2023   Prob (F-statistic):           8.50e-50
Time:                        14:49:47   Log-Likelihood:                -527.10
No. Observations:                 634   AIC:                             1456.
Df Residuals:                     433   BIC:                             2351.
Df Model:                         200                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        2.0000      0.672      2.974   

This is a linear regression analysis that predicts the round number (dependent variable) based on the time at which the round finished (independent variable). The R-squared value of 0.717 suggests that approximately 71.7% of the variance in round number can be explained by the time at which the round finished. The Adj. R-squared value of 0.587 suggests that the model may not fit the data as well as it could, as it is less than the R-squared value. The F-statistic value of 5.494 is a measure of the overall significance of the regression and indicates that the model is significant at the p<0.05 level. The Prob (F-statistic) value of 8.50e-50 is the probability of getting an F-statistic as large as the one calculated by chance, and since it is very small, we can reject the null hypothesis that the model is not significant.